http/vibe/http/common.d

```diff
@@ -336,13 +337,222 @@ class HTTPStatusException : Exception {
+		ret.headers["Content-Disposition"] = "file; filename=\"" ~ filename ~ "\"";
+		ret.headers["Content-Type"] = content_type;
+		ret.content = stream;
+		return ret;
+	}
+
+	static MultiPartBodyPart multipleFilesPart(string filename, string content_type, string content)
+	{
+		auto ret = new MultiPartBodyPart();
+		ret.headers["Content-Disposition"] = "file; filename=\"" ~ filename ~ "\"";
+		ret.headers["Content-Type"] = content_type;
+		ret.content = createMemoryStream((() @trusted => cast(ubyte[]) content)(), false);
+		return ret;
+	}
+
+	static MultiPartBodyPart multipleFilesPart(string filename, string content_type, ubyte[] content)
+	{
+		auto ret = new MultiPartBodyPart();
+		ret.headers["Content-Disposition"] = "file; filename=\"" ~ filename ~ "\"";
+		ret.headers["Content-Transfer-Encoding"] = "binary";
+		ret.headers["Content-Type"] = content_type;
+		ret.content = createMemoryStream(content, false);
+		return ret;
+	}
+
+	static MultiPartBodyPart multipleFiles(string name, MultiPart multipart)
+	{
+		import vibe.stream.memory : createMemoryOutputStream;
+
+		auto ret = new MultiPartBodyPart();
+		string boundary = randomMultipartBoundary;
+		ret.headers["Content-Disposition"] = "form-data; name=\"" ~ name ~ "\"";
+		ret.headers["Content-Type"] = "multipart/mixed; boundary=\"" ~ boundary ~ "\"";
+		auto stream = createMemoryOutputStream();
+		multipart.write(boundary, stream);
+		ret.content = createMemoryStream(stream.data, false);
+		return ret;
+	}
+}
+
+final class MultiPart
+{
+	// https://www.w3.org/Protocols/rfc1341/7_2_Multipart.html
+
+	string contentType = "multipart/form-data";
+	string preamble, epilogue;
+
+	MultiPartBodyPart[] parts;
+
+	@safe:
+
+	size_t length(string boundary) const
+	{
+		if (!parts.length)
+			return 0;
+
+		size_t length;
+		if (preamble.length)
+			length += preamble.length + 2;
+		length += boundary.length + 2; // --boundary
+		foreach (part; parts) {
+			length += 8 + boundary.length; // \r\n * 3 + boundary (+2)
+			foreach (k, v; part.headers)
+				length += 4 + k.length + v.length; // ": \r\n"
+			auto randomStream = cast(const(RandomAccessStream)) part.content;
+			if (randomStream)
+				length += randomStream.size;
+			else
+				return 0; // undetectable
+		}
+		length += 4;
+		if (epilogue.length)
+			length += epilogue.length + 2;
+		return length;
+	}
+
+	void write(T)(string boundary, T output)
+		if (isOutputStream!T)
+	{
+		import vibe.core.stream : pipe;
+
+		if (!parts.length)
+			return;
+
+		boundary = "--" ~ boundary;
+		if (preamble.length)
+			output.write(preamble ~ "\r\n");
+		output.write(boundary);
+		foreach (part; parts) {
+			output.write("\r\n");
+			foreach (k, v; part.headers)
+				output.write(k ~ ": " ~ v ~ "\r\n");
+			output.write("\r\n");
+			pipe(part.content, output);
+			output.write("\r\n");
+			output.write(boundary);
+		}
+		output.write("--\r\n");
+		if (epilogue.length)
+			output.write(epilogue ~ "\r\n");
+	}
+}
+
+string randomMultipartBoundary()
```

Comment:
I'd keep this private for now, since there is no obvious public use for this and it would mean that the exact API isn't critical.

Reply:
this is a public API now and is fairly critical if you want to nest multiparts. Because MultiPart is now in vibe.inet it also doesn't have access to crypto which is now used in this API